linera-views/src/indexed_db.rs

```diff
@@ -0,0 +1,359 @@
+        let common_config = CommonStoreConfig {
+            max_concurrent_queries: None,
+            max_stream_queries,
+            cache_size: 1000,
+        };
+        Self { common_config }
+    }
+}
+
+/// The number of streams for the test
+pub const TEST_INDEX_DB_MAX_STREAM_QUERIES: usize = 10;
+
+const DATABASE_NAME: &str = "linera-views";
+
+/// The data is serialized in memory just like for RocksDB / DynamoDB
+/// The analog of the database is the BTreeMap
+pub type IndexedDbStoreMap = BTreeMap<Vec<u8>, Vec<u8>>;
+
+/// A virtual DB client where data are persisted in memory.
+pub struct IndexedDbStore {
+    /// The database used for storing the data.
+    pub database: IdbDatabase,
+    /// The object store name used for storing the data.
+    pub object_store_name: String,
+    /// The maximum number of queries used for the stream.
+    pub max_stream_queries: usize,
+}
+
+impl IndexedDbStore {
+    fn with_object_store<R>(
+        &self,
+        f: impl FnOnce(IdbObjectStore) -> R,
+    ) -> Result<R, IndexedDbContextError> {
+        let transaction = self.database.transaction_on_one(&self.object_store_name)?;
+        let object_store = transaction.object_store(&self.object_store_name)?;
+        Ok(f(object_store))
+    }
+}
+
+fn prefix_to_range(prefix: &[u8]) -> Result<web_sys::IdbKeyRange, wasm_bindgen::JsValue> {
+    let lower = js_sys::Uint8Array::from(prefix);
+    if let Some(upper) = get_upper_bound_option(prefix) {
+        let upper = js_sys::Uint8Array::from(&upper[..]);
+        web_sys::IdbKeyRange::bound_with_lower_open_and_upper_open(
+            &lower.into(),
+            &upper.into(),
+            false,
+            true,
+        )
+    } else {
+        web_sys::IdbKeyRange::lower_bound(&lower.into())
+    }
+}
+
+impl LocalReadableKeyValueStore<IndexedDbContextError> for IndexedDbStore {
+    const MAX_KEY_SIZE: usize = usize::MAX;
+    type Keys = Vec<Vec<u8>>;
+    type KeyValues = Vec<(Vec<u8>, Vec<u8>)>;
+
+    fn max_stream_queries(&self) -> usize {
+        self.max_stream_queries
+    }
+
+    async fn read_value_bytes(&self, key: &[u8]) -> Result<Option<Vec<u8>>, IndexedDbContextError> {
+        let key = js_sys::Uint8Array::from(key);
+        let value = self.with_object_store(|o| o.get(&key))??.await?;
+        Ok(value.map(|v| js_sys::Uint8Array::new(&v).to_vec()))
+    }
+
+    async fn contains_key(&self, key: &[u8]) -> Result<bool, IndexedDbContextError> {
+        let key = js_sys::Uint8Array::from(key);
+        let count = self.with_object_store(|o| o.count_with_key(&key))??.await?;
+        assert!(count < 2);
+        Ok(count == 1)
+    }
+
+    async fn read_multi_values_bytes(
+        &self,
+        keys: Vec<Vec<u8>>,
+    ) -> Result<Vec<Option<Vec<u8>>>, IndexedDbContextError> {
+        let mut values = Vec::with_capacity(keys.len());
+        for key in keys {
+            values.push(self.read_value_bytes(&key).await?);
+        }
```

Comment:
No join! here because everything is serial I guess.

Reply:
I didn't use `join!` here because it would make the order of the values non-deterministic.  I suppose I could also `join!` to get key/value pairs and then sort by key.

Reply:
`futures::future::join_all` doesn't change the order, I think.

Reply:
Nice!  Done.